File: java/Stack/src/main/java/stack/LinkedListStack.java
package stack;

import java.util.Iterator;
import java.util.ListIterator;

public class LinkedListStack<I>  implements Stack<I>{

    class Node {
        I data;
        Node next;
        Node(I data) { this.data =  data;}
    }

    Node first;
    int size = 0;

    public void push(I item) {
        Node oldNode = first;
        first = new Node(item);
        first.next = oldNode;
        size++;
    }

    public I pop() {
        I myVal = first.data;
        first = first.next;
        size--;
        return myVal;
    }

    public boolean isEmpty() {
        return first == null;
    }

    public int size() {
        return size;
    }


    // Additional method for provide more convenience

    public Iterator iterator() {
        return new ListIterator() ;
    }

    private class ListIterator implements Iterator<I> {

        private Node current = first ;

        public boolean hasNext() { return current != null; }

        public I next() {
            I item = current.data;
            current = current.next;
            return item;
        }
    }
}

File: cxx/Stack/LinkedListStack.cpp
#include <iostream>

#include "Stack.hpp"

using namespace std;
using namespace Stack;

namespace Stack {

// -------------------- LinkedList Definition -----------------------------------
template<class T>
class LinkedListStack  : public Stack<T>{

private:
    template<class T>
    class Link {
        public :
            T data;
            Link *next;
    };

    Link<T> *stack;
public:
    LinkedListStack();
    ~LinkedListStack();
    virtual void push(T item);
    virtual T pop();
    virtual bool isEmpty();
    virtual int size();
};

template<class T>
LinkedListStack<T>::LinkedListStack() {
    this->stack = new Link<T>();
};

template<class T>
LinkedListStack<T>::~LinkedListStack(){

};

template<class T>
void LinkedListStack<T>::push(T item) {
    cout << "Item ->" << item << endl ;
};

template<class T>
T LinkedListStack<T>::pop(){
    T popper = this->stack->data;
    return popper;
};

template<class T>
bool LinkedListStack<T>::isEmpty(){
    return false;
};

template<class T>
int LinkedListStack<T>::size() {
    return 1;
};
// ------------------------------------------------------------------------------

}

